### PR TITLE
Create code action for propose similar enum value

### LINKED ISSUE #130: CodeAction to fix value property by proposing similar value
Like https://github.com/redhat-developer/quarkus-ls/issues/80 for property name, it should be good to have the same feature for value when value is an enumeration.

### PR DESCRIPTION
**Pull request by xorye**
Fixes #130

Gif showing codeaction

Signed-off-by: David Kwon <dakwon@redhat.com>

### CODE REVIEW

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/commons/QuarkusProjectInfo.java [hunk 1]
```diff
@@ -64,6 +64,22 @@ public void setClasspathKind(ClasspathKind classpathKind) {
...
 	}
 
+	/**
+	 * Returns a <code>ExtendedConfigDescriptionBuildItem</code> representing the
+	 * <code>propertyName</code> property
+	 * @param propertyName the property name of interest
+	 * @return <code>ExtendedConfigDescriptionBuildItem</code> representing the
+	 * <code>propertyName</code> property
+	 */
+	public ExtendedConfigDescriptionBuildItem getPropertyFromName(String propertyName) {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/commons/QuarkusProjectInfo.java]: Remove this method to use `QuarkusPropertiesUtils#getProperty(String propertyName, QuarkusProjectInfo info)` which manages search of mapped property.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java [hunk 2]
```diff
@@ -111,6 +118,50 @@ private void doCodeActionsForUnknown(Diagnostic diagnostic, PropertiesModel docu
...
+	 */
+	private void doCodeActionsForUnknownEnumValue(Diagnostic diagnostic, PropertiesModel document,
+			QuarkusProjectInfo projectInfo, ValuesRulesManager valuesRulesManager, List<CodeAction> codeActions) {
+		try {
+			PropertyValue propertyValue = (PropertyValue) document.findNodeAt(diagnostic.getRange().getStart());
+			PropertyKey propertyKey = ((Property) propertyValue.getParent()).getKey();
+			String value = propertyValue.getValue();
+			String propertyName = propertyKey.getPropertyName();
+
+			ExtendedConfigDescriptionBuildItem metaProperty = projectInfo.getPropertyFromName(propertyName);
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java]: Replace with:

```java
ExtendedConfigDescriptionBuildItem metaProperty = QuarkusPropertiesUtils.getProperty(propertyName, projectInfo)
```

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java [hunk 3]
```diff
@@ -111,6 +118,50 @@ private void doCodeActionsForUnknown(Diagnostic diagnostic, PropertiesModel docu
...
+			if (metaProperty == null) {
+				return;
+			}
+
+			List<EnumItem> enums = metaProperty.getEnums();
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java]: In QuarkusCompletions, you have a private static method

```java
 Collection<EnumItem> getEnums(ExtendedConfigDescriptionBuildItem property, PropertiesModel model,
			ValuesRulesManager valuesRulesManager) 
```
which does the same thing (nut manages boolean type too).

I suggest you to move this private static method from QuarkusCompletion in the QuarkusPropertiesUtils and you consume it here. 
Replace this enums collection by creating a new method utility `Collection<EnumItem> getEnums(ExtendedConfigDescriptionBuildItem property, PropertiesModel model,	ValuesRulesManager valuesRulesManager)` You can find this method in

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java [hunk 4]
```diff
@@ -111,6 +118,50 @@ private void doCodeActionsForUnknown(Diagnostic diagnostic, PropertiesModel docu
...
+			if (enums == null || enums.isEmpty()) {
+				enums = valuesRulesManager.getValues(metaProperty, document);
+				if (enums == null || enums.isEmpty()) {
+					return;
+				}
+			}
+
+			for (EnumItem e : enums) {
+				if (isSimilar(e.getName(), value)) {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java]: In LSP4XML we have a current PR which does the same thing (but with XML Schema). The idea is to provide a code actions for similar names (like you have done). And if you have none similar name, you provide a code action per enum item by saying 'Replace with '

See https://github.com/angelozerr/lsp4xml/blob/655e17c197011cf41fe1880c54a861f2c1ef00c2/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/codeactions/cvc_complex_type_2_4_aCodeAction.java#L76

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesCodeActionsTest.java [hunk 5]
```diff
@@ -42,4 +42,29 @@ public void codeActionsForUnknownProperties() throws BadLocationException {
...
+		String value = "quarkus.log.syslog.async.overflow=BLACK";
+
+		Diagnostic d = d(0, 34, 39, "Invalid enum value: 'BLACK' is invalid for type org.jboss.logmanager.handlers.AsyncHandler$OverflowAction", DiagnosticSeverity.Error,
+				ValidationType.value);
+		testDiagnosticsFor(value, d);
+
+		testCodeActionsFor(value, d,
+				ca("Did you mean 'BLOCK'?", te(0, 34, 0, 39, "BLOCK"), d));
+	};
+
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesCodeActionsTest.java]: Add test with `Replace with` code action.

**Comment by xorye** [comment]
Thank you for the great feedback, I've made the changes

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesCodeActionsTest.java [hunk 6]
```diff
@@ -37,9 +37,56 @@ public void codeActionsForUnknownProperties() throws BadLocationException {
...
+		Diagnostic d = d(0, 34, 47,
+				"Invalid enum value: 'unknown-value' is invalid for type org.jboss.logmanager.handlers.AsyncHandler$OverflowAction",
+				DiagnosticSeverity.Error,
+				ValidationType.value);
+
+		testDiagnosticsFor(value, d);
+		testCodeActionsFor(value, d,
+				ca("Replace with 'BLOCK'?", te(0, 34, 0, 47, "BLOCK"), d),
+				ca("Replace with 'DISCARD'?", te(0, 34, 0, 47, "DISCARD"), d));
+	};
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesCodeActionsTest.java]: Add a test with the mapped property `quarkus.log.category."org.acme".level=WARNIN`

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java [hunk 7]
```diff
@@ -111,6 +119,64 @@ private void doCodeActionsForUnknown(Diagnostic diagnostic, PropertiesModel docu
 		}
 	}
 
+	/**
+	 * Create code action fo
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java]: Fix javadoc

**Comment by angelozerr** [comment]
The PR works very good, but similar name works if I write

`quarkus.log.level=WARNIN`

but if I write

`quarkus.log.level=WARNI`

It doesn't work (we have replace code action with all values instead of having similar code action with WARNING). 

In this case, 
`int threshold = Math.round(MAX_DISTANCE_DIFF_RATIO * reference.length());` is equals to 0. @fbricon have you an idea if we could manage the case with WARNI

WARNI

**Comment by angelozerr** [comment]
I'm OK to merge this PR, but I wonder if `quarkus.log.level=WARNI` should work (I mean provide a code action which says `Did you mean 'WARNING' ?`) but it's a bug (or not?) from isSimilar method.

Perhaps we could merge this PR and fix this case (if it's possible) in an another issue. @fbricon what do you think?

**Comment by fbricon** [comment]
Regardless of the levenshtein distance, there's only one enum that starts with the given value, it should be proposed

**Comment by angelozerr** [comment]
> Regardless of the levenshtein distance, if there's only one enum that starts with the given value, it should be proposed

I agree with you, @xorye is there any chance that you work on this levenshtein issue in this PR? @fbricon if you have some idea how to fix it, you are welcome!

**Comment by angelozerr** [comment]
> Regardless of the levenshtein distance

Ok I understand more your idea, we must check starts with at first in the isSimilar method. I think we should do that but only for value and not for property name.

**Comment by xorye** [comment]
>  I think we should do that but only for value and not for property name.

Thanks for the feedback, I have it implemented like this: 

https://github.com/redhat-developer/quarkus-ls/blob/f5a3a219a4656c08e775582d32114c739ca1d61d/quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java#L270-L272

I also added tests for this case

**Comment by angelozerr** [comment]
@xorye it works like a charm, great job!

**Pull request merged**